gravitee-node-secrets/gravitee-node-secrets-service/src/main/java/io/gravitee/node/secrets/service/AbstractSecretProviderDispatcher.java

```diff
@@ -0,0 +1,114 @@
+    private final SecretProviderPluginManager secretProviderPluginManager;
+
+    private final Map<String, SecretProvider> secretProviders = new HashMap<>();
+
+    protected AbstractSecretProviderDispatcher(SecretProviderPluginManager secretProviderPluginManager) {
+        this.secretProviderPluginManager = secretProviderPluginManager;
+    }
+
+    protected final void createAndRegister(String id) {
+        try {
+            final SecretProviderPlugin<?, ?> secretProviderPlugin = secretProviderPluginManager.get(id);
+            final Class<? extends SecretManagerConfiguration> configurationClass = secretProviderPlugin.configuration();
+            final SecretProviderFactory<SecretManagerConfiguration> factory = secretProviderPluginManager.getFactoryById(id);
+            if (configurationClass != null && factory != null) {
+                // read the config using the plugin class loader
+                SecretManagerConfiguration config =
+                    this.readConfiguration(id, factory.getClass().getClassLoader().loadClass(configurationClass.getName()));
+                // register and start
+                secretProviders.put(id, factory.create(config).start());
+            } else {
+                throw new SecretProviderNotFoundException(SECRET_PROVIDER_NOT_FOUND_FOR_ID.formatted(id));
+            }
+        } catch (ClassNotFoundException e) {
+            throw new IllegalArgumentException("cannot load plugin %s properly: ".formatted(id));
+        }
+    }
+
+    @Override
+    public Maybe<SecretMap> resolve(SecretMount secretMount) throws SecretProviderNotFoundException, SecretManagerException {
+        return secretProviders
+            .getOrDefault(secretMount.provider(), new ErrorSecretProvider())
+            .resolve(secretMount)
+            .subscribeOn(Schedulers.io());
+    }
+
+    @Override
+    public Maybe<Secret> resolveKey(SecretMount secretMount) throws SecretProviderNotFoundException, SecretManagerException {
+        if (secretMount.isKeyEmpty()) {
+            return Maybe.error(new IllegalArgumentException("cannot request secret key, no key provided"));
+        }
+        return resolve(secretMount).flatMap(secretMap -> Maybe.fromOptional(secretMap.getSecret(secretMount)));
+    }
+
+    @Override
+    public Flowable<SecretMap> watch(SecretMount secretMount, SecretEvent.Type... events) {
+        final SecretProvider provider = secretProviders.getOrDefault(secretMount.provider(), new ErrorSecretProvider());
+        return provider
+            .watch(secretMount)
+            .filter(secretEvent -> events == null || events.length == 0 || Arrays.asList(events).contains(secretEvent.type()))
+            .subscribeOn(Schedulers.io())
+            .map(SecretEvent::secretMap)
+            .doFinally(provider::stop);
+    }
+
+    @Override
+    public Flowable<Secret> watchKey(SecretMount secretMount, SecretEvent.Type... events) {
```

Comment:
question: events is not used, did you forget it in the implementation or does the method signature is not valid anymore ?

Reply:
you are right, I move the filtering responsibility to the dispatcher as plugin are meant to be as dumb as possible